src/main/scala/org/broadinstitute/dsde/agora/server/webservice/AgoraService.scala

```diff
@@ -0,0 +1,85 @@
+
+package org.broadinstitute.dsde.agora.server.webservice
+
+import org.broadinstitute.dsde.agora.server.model.{AgoraEntity, AgoraEntityProjection, AgoraEntityType}
+import org.broadinstitute.dsde.agora.server.webservice.util.ServiceMessages
+import org.broadinstitute.dsde.agora.server.webservice.validation.{AgoraValidation, AgoraValidationRejection}
+import org.joda.time.DateTime
+import spray.routing.HttpService
+
+trait AgoraService extends HttpService with PerRequestCreator with AgoraDirectives {
+  this: ServiceHandlerProps =>
+
+  private implicit val executionContext = actorRefFactory.dispatcher
+
+  import org.broadinstitute.dsde.agora.server.model.AgoraApiJsonSupport._
+  import spray.httpx.SprayJsonSupport._
+
+  def path: String
+
+  def routes = queryByNamespaceNameSnapshotIdRoute ~ queryRoute ~ postRoute
+
+  def queryByNamespaceNameSnapshotIdRoute =
+    path(path / Segment / Segment / Segment) { (namespace, name, snapshotId) =>
+      get {
+        requestContext =>
+          perRequest(requestContext, queryHandlerProps, ServiceMessages.QueryByNamespaceNameSnapshotId(requestContext, namespace, name, snapshotId.toInt,
+            AgoraEntityType.byPath(path)))
+      }
+    }
+
+  def queryRoute =
+    path(path) {
+      get {
+        parameters(
+          "namespace".?,
+          "name".?,
+          "snapshotId".as[Int].?,
+          "synopsis".?,
+          "documentation".?,
+          "owner".?,
+          "createDate".as[DateTime].?,
+          "payload".?,
+          "url".?,
+          "entityType".as[AgoraEntityType.EntityType].?
```

Comment:
Why can we search by entity type in metadata?  Wouldn't we be calling GET with parameters on different endpoints?  What would it mean if I searched /configurations with entityType=methods?